File: exams/ExamFunctions.vb
Module ExamFunctions



    Private Sub copyValues(ByVal grades As Dictionary(Of String, List(Of Tuple(Of String, Int16))))
        subjectGrading2.Clear()
        subjectGrading2 = grades
    End Sub

    Public Function getGradingScheme() As Boolean
        Dim schemeAvailable As Boolean = False
        Dim query As String = String.Empty
        Dim gradess As String() = {"A", "A-", "B+", "B", "B-", "C+", "C", "C-", "D+", "D", "D-", "E"}
        Dim gradeValues As New List(Of Tuple(Of String, Int16))
        Dim saveGradeValue As New List(Of Tuple(Of String, Int16))


        If String.IsNullOrEmpty(gradingType) Then
            gradingType = String.Empty
        End If

        If gradingType = "SubjectBased" Then

            subjectGrading.Clear()

            For Each s As String In schoolSubjects
                query = "SELECT * FROM s_grading WHERE (Class='" & selectedClass & "' AND Subject='" & get_subject_name(s) & "' AND term='" & selectedTerm & "' AND year='" & selectedYear & "');"
                If qread(query, 1) Then
                    If dbreader1.RecordsAffected > 0 Then
                        schemeAvailable = True
                        While dbreader1.Read
                            For Each g As String In gradess
                                gradeValues.Add(New Tuple(Of String, Int16)(g, dbreader1(g)))
                            Next
                            ' test(s, gradeValues)
                            subjectGrading.Add(s, gradeValues)
                            gradeValues.Clear()
                        End While
                    End If
                End If
            Next

        Else

            classGrading.Clear()

            query = "SELECT * FROM grading WHERE (Class='" & selectedClass & "' AND term='" & selectedTerm & "' AND year='" & selectedYear & "');"
            If qread(query, 1) Then
                If dbreader1.RecordsAffected > 0 Then
                    schemeAvailable = True
                    While dbreader1.Read
                        For Each g As String In gradess
                            classGrading.Add(New Tuple(Of String, Int16)(g, dbreader1(g)))
                        Next
                    End While
                End If
            End If
        End If

        Return schemeAvailable
    End Function

    Public Sub initialzeGradePoints()

        GradeToPoint.Clear()

        GradeToPoint.Add("A", 12)
        GradeToPoint.Add("A-", 11)
        GradeToPoint.Add("B+", 10)
        GradeToPoint.Add("B", 9)
        GradeToPoint.Add("B-", 8)
        GradeToPoint.Add("C+", 7)
        GradeToPoint.Add("C", 6)
        GradeToPoint.Add("C-", 5)
        GradeToPoint.Add("D+", 4)
        GradeToPoint.Add("D", 3)
        GradeToPoint.Add("D-", 2)
        GradeToPoint.Add("E", 1)
        GradeToPoint.Add("Y", 0)
        GradeToPoint.Add("X", 0)

    End Sub

    Public Function convertMarksToGrade(ByVal marks As Int16, ByVal gradingType As String, Optional ByVal subjectParm As String = "") As String
        Dim studGrade As String = String.Empty

        If marks = 0 Then
            studGrade = "Y"
        ElseIf gradingType = "SubjectBased" Then
            'Dim subjectGrading As New Dictionary(Of String, List(Of Tuple(Of String, Int16)))

            Dim subjGrade As New List(Of Tuple(Of String, Int16))
            For Each pair In subjectGrading
                If pair.Key = subjectParm Then
                    subjGrade = pair.Value
                End If
            Next


            If marks < subjGrade(subjGrade.Count - 1).Item2 Then
                    studGrade = "E"
                Else
                    For i = 0 To subjGrade.Count - 1
                        If marks >= subjGrade(i).Item2 Then
                            studGrade = subjGrade(i).Item1
                            Exit For
                        End If
                    Next
                End If
            Else
                ' Dim classGrading As New List(Of Tuple(Of String, Int16))
                If marks < classGrading(classGrading.Count - 1).Item2 Then
                studGrade = "E"
            Else
                For i = 0 To classGrading.Count - 1
                    If marks >= classGrading(i).Item2 Then
                        studGrade = classGrading(i).Item1
                        Exit For
                    End If
                Next
            End If
        End If

        Return studGrade
    End Function

    Public Function convertGradeToPoints(ByVal grade As String) As Int16

        Dim points As Int16
        points = GradeToPoint(grade)
        Return points

    End Function

End Module

File: exams/ExamFunctions.cs
using System;
using System.Collections.Generic;
using Microsoft.VisualBasic.CompilerServices;

namespace exams
{
    static class ExamFunctions
    {
        private static void copyValues(Dictionary<string, List<Tuple<string, short>>> grades)
        {
            publicSubsNFunctions.subjectGrading2.Clear();
            publicSubsNFunctions.subjectGrading2 = grades;
        }

        public static bool getGradingScheme()
        {
            bool schemeAvailable = false;
            string query = string.Empty;
            var gradess = new[] { "A", "A-", "B+", "B", "B-", "C+", "C", "C-", "D+", "D", "D-", "E" };
            var gradeValues = new List<Tuple<string, short>>();
            var saveGradeValue = new List<Tuple<string, short>>();
            if (string.IsNullOrEmpty(publicSubsNFunctions.gradingType))
            {
                publicSubsNFunctions.gradingType = string.Empty;
            }

            if (publicSubsNFunctions.gradingType == "SubjectBased")
            {
                publicSubsNFunctions.subjectGrading.Clear();
                foreach (string s in publicSubsNFunctions.schoolSubjects)
                {
                    query = Conversions.ToString(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject("SELECT * FROM s_grading WHERE (Class='" + publicSubsNFunctions.selectedClass + "' AND Subject='", publicSubsNFunctions.get_subject_name(s)), "' AND term='"), publicSubsNFunctions.selectedTerm), "' AND year='"), publicSubsNFunctions.selectedYear), "');"));
                    if (publicSubsNFunctions.qread(ref query, 1))
                    {
                        if (publicSubsNFunctions.dbreader1.RecordsAffected > 0)
                        {
                            schemeAvailable = true;
                            while (publicSubsNFunctions.dbreader1.Read())
                            {
                                foreach (string g in gradess)
                                    gradeValues.Add(new Tuple<string, short>(g, Conversions.ToShort(publicSubsNFunctions.dbreader1[g])));
                                // test(s, gradeValues)
                                publicSubsNFunctions.subjectGrading.Add(s, gradeValues);
                                gradeValues.Clear();
                            }
                        }
                    }
                }
            }
            else
            {
                publicSubsNFunctions.classGrading.Clear();
                query = "SELECT * FROM grading WHERE (Class='" + publicSubsNFunctions.selectedClass + "' AND term='" + publicSubsNFunctions.selectedTerm + "' AND year='" + publicSubsNFunctions.selectedYear + "');";
                if (publicSubsNFunctions.qread(ref query, 1))
                {
                    if (publicSubsNFunctions.dbreader1.RecordsAffected > 0)
                    {
                        schemeAvailable = true;
                        while (publicSubsNFunctions.dbreader1.Read())
                        {
                            foreach (string g in gradess)
                                publicSubsNFunctions.classGrading.Add(new Tuple<string, short>(g, Conversions.ToShort(publicSubsNFunctions.dbreader1[g])));
                        }
                    }
                }
            }

            return schemeAvailable;
        }

        public static void initialzeGradePoints()
        {
            publicSubsNFunctions.GradeToPoint.Clear();
            publicSubsNFunctions.GradeToPoint.Add("A", 12);
            publicSubsNFunctions.GradeToPoint.Add("A-", 11);
            publicSubsNFunctions.GradeToPoint.Add("B+", 10);
            publicSubsNFunctions.GradeToPoint.Add("B", 9);
            publicSubsNFunctions.GradeToPoint.Add("B-", 8);
            publicSubsNFunctions.GradeToPoint.Add("C+", 7);
            publicSubsNFunctions.GradeToPoint.Add("C", 6);
            publicSubsNFunctions.GradeToPoint.Add("C-", 5);
            publicSubsNFunctions.GradeToPoint.Add("D+", 4);
            publicSubsNFunctions.GradeToPoint.Add("D", 3);
            publicSubsNFunctions.GradeToPoint.Add("D-", 2);
            publicSubsNFunctions.GradeToPoint.Add("E", 1);
            publicSubsNFunctions.GradeToPoint.Add("Y", 0);
            publicSubsNFunctions.GradeToPoint.Add("X", 0);
        }

        public static string convertMarksToGrade(short marks, string gradingType, string subjectParm = "")
        {
            string studGrade = string.Empty;
            if (marks == 0)
            {
                studGrade = "Y";
            }
            else if (gradingType == "SubjectBased")
            {
                // Dim subjectGrading As New Dictionary(Of String, List(Of Tuple(Of String, Int16)))

                var subjGrade = new List<Tuple<string, short>>();
                foreach (var pair in publicSubsNFunctions.subjectGrading)
                {
                    if ((pair.Key ?? "") == (subjectParm ?? ""))
                    {
                        subjGrade = pair.Value;
                    }
                }

                if (marks < subjGrade[subjGrade.Count - 1].Item2)
                {
                    studGrade = "E";
                }
                else
                {
                    for (int i = 0, loopTo1 = subjGrade.Count - 1; i <= loopTo1; i++)
                    {
                        if (marks >= subjGrade[i].Item2)
                        {
                            studGrade = subjGrade[i].Item1;
                            break;
                        }
                    }
                }
            }
            // Dim classGrading As New List(Of Tuple(Of String, Int16))
            else if (marks < publicSubsNFunctions.classGrading[publicSubsNFunctions.classGrading.Count - 1].Item2)
            {
                studGrade = "E";
            }
            else
            {
                for (int i = 0, loopTo = publicSubsNFunctions.classGrading.Count - 1; i <= loopTo; i++)
                {
                    if (marks >= publicSubsNFunctions.classGrading[i].Item2)
                    {
                        studGrade = publicSubsNFunctions.classGrading[i].Item1;
                        break;
                    }
                }
            }

            return studGrade;
        }

        public static short convertGradeToPoints(string grade)
        {
            short points;
            points = publicSubsNFunctions.GradeToPoint[grade];
            return points;
        }
    }
}